Fix join room failures on large lobbies

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/11294

I looked at potentially changing this to use the locking flow I'm using server side but due to the number of usages (and the majority already correctly scheduled) I went in this direction instead. We can reassess at a future point in time if locking is deemed what we want.

This also fixes user population running single file and bypassing the `UserLookupCache` batching.

peppy committed:
Fix user population happening in single file

peppy committed:
Fix non-safe thread access to room users on room join

bdach commented:
Looking at the failing tests here - for what it's worth, the first one I'm looking at, I actually have no idea how it ever worked.

bdach commented:
Never mind, most likely all the test failures are due to that `getRoomUsers()` logic. The schedule doesn't manage to run in time before the reset event times out, joining the room fails because a null user list is return, at which point the client leaves the room and nulls it out, leaving behind a quite broken state that tests definitely cannot run in.

peppy commented:
Shall we just increase the timeout and call it a day? It was arbitrary anyway.

bdach commented:
I kind of want to see why the schedule runs so late first.

bdach committed:
Run user list copy inline if possible

`getRoomUsers()` was not safe to call from the update thread, as
evidenced by the test failures. This was due to the fact that the added
reset event could never actually be set from within the method, as the
wait was blocking the scheduled set from ever proceeding.

Resolve by allowing the scheduled copy & set to run inline if on the
update thread already.

bdach committed:
Allow all StatefulMultiplayerClient schedules to run inline

Fixes test failures due to not allowing to do so, therefore inverting
execution order in some cases - for example, calling

        JoinRoom(room);
        LeaveRoom();

on the update thread would invert execution order due to the first being
unscheduled but the second being scheduled.

bdach commented:
Pushed out changes that should hopefully allow the tests to pass (at least they did locally, and I checked a few times). Make sure you're OK with them, but there should be no surprises.

bdach commented:
Well, changes certainly seem to have helped.

Not a 100% on this general method, locking would be preferable as mentioned, but let's take it step by step for now...

peppy commented:
Changes look fine.
